### DISCUSSION TITLE
How to best propagate constant config information between units?

### DISCUSSION THREAD

**Discussion by egieske** [riscv-software-src/riscv-perf-model#99] [Q&A] [answered]
How should I communicate the ICache line size from the ICache unit to the Fetch unit? I want to prevent fetching past the end of the cache line. Ports are for dynamic inter-unit communication. Can one unit access another's parameters? If so, is there an example in Olympia and will the accessed parameters include the default values?

**Comment by arupc** [comment, marked as answer]
One solution (just as a workaround) would to use two different params, one for fetch unit, another Icache unit and user being responsible such that values of those parameters are equal. This is just to test rest of the code.

The other more permanent solution could be to specifiy icache line size as a global simulation argument, just like workload/trace is specified.

I agree better solution would for one sparta unit to access the value of the parameter for other sparta unit. @klingaard is that possible?

**Comment by ghost** [comment]
There are a couple of avenues.  For one, if you are designing the Fetch block to _always_ have the icache as a child of the Fetch unit, you can (in the constructor of Fetch) grab the parameters of the icache block and query the size:
```
Fetch::Fetch(sparta::TreeNode* node, const FetchParameterSet* p) : ...
{
    auto icache_node = node->getChild("icache"); // or whatever you called it
    sparta_assert(nullptr != icache_node);
    auto icache_size = icache_node->getParameterSet()->getParameter("icache_size")->getParameterValueAs<uint32_t>();
}
```
The way I _prefer_ the icache size to be defined and retrieved is to do something similar to what Arup says and make the parameter a core extension.

[To this file](https://github.com/riscv-software-src/riscv-perf-model/blob/82504f6f68374d4b335d9de3d5d3890fe55e9838/core/CPUTopology.hpp#L154) (and this function), add this code:
```
            if(nullptr == ps->getParameter("icache_size", false)) {
                icache_size_.reset(new sparta::Parameter<uint32_t>(
                                          "icache_size", 128,
                                          "ICache size in KB", ps));
            }
```
and add the `icache_size_` to the `private` section of the class:
```
        std::unique_ptr<sparta::Parameter<uint32_t>>    icache_size_;
```

You can get the ICache size in both the Fetch Unit and the ICache Unit via a helper method you add to [CoreUtils.hpp](https://github.com/riscv-software-src/riscv-perf-model/blob/82504f6f68374d4b335d9de3d5d3890fe55e9838/core/CoreUtils.hpp#L24).
(copy the concept in `getExecutionTopology`).
